proposals/kapp-controller/004-dependency-resolution/README.md

```diff
@@ -0,0 +1,619 @@
+        kind: MainApi
+      conditions:
+      - type: ExistingGVK
+
+    - name: dep3
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1"
+      packageInstallRef: 
+        name: name: dep-pkgi.this-is-a-random-string
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: dep4
+      wantedPackageRef:
+        refName: pkg4.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef: 
+        name: dep-pkgi.this-is-another-random-string
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: microservice-v1
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: other-pkg.pkg5.test.carvel.dev
+        namespace: ns1
+      conditions:
+      - type: FoundPackageInstall
+
+    - name: microservice-v2
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: pkg-demo.microservice-v2
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: optional-dependency
+      wantedPackageRef:
+        refName: pkg6.test.carvel.dev
+        versionSelection:
+          constraints: 0.6.0
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-wrong-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1" # one on the cluster is 0.1.1
+      packageInstallRef:
+        name: pkg-demo.pkg1.test.carvel.dev
+      conditions:
+      - type: FoundPackageInstall
+      - type: FoundPackageInstallRequiresDifferentVersion
+
+    - name: dep-with-unknown-package
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-package-with-high-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: ">4.0.0" # cant calculate exact version in this system
+      conditions:
+      - type: UnknownPackageVersion
+
+```
+
+Example of a PackageInstall that is automatically created
+```yaml
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
+spec:
+    serviceAccountName: default-ns-sa
+  packageRef:
+    refName: pkg3.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+  values:
+    secretRef:
+      name: pkg3.dev-values
+...
+```
+
+#### Dependency resolution
+The dependency resolution will be done one PackageInstall at a time, basically each PackageInstall will only create the PackageInstalls that it directly depends on.
```

Comment:
When there is a new dependency that gets identified for a specific `PackageInstall`, can the `PackageInstall` be updated with that dep after it's been installed in the cluster?

Reply:
Yes! kapp-controller will try to reconcile PackageInstalls after every `syncPeriod` duration, so in the meanwhile if there have been any new dependencies added to the package, then those would be installed in the next reconciliation.